design-proposals/volume-migration/volume-migration.md

```diff
@@ -0,0 +1,466 @@
+}
+
+type VolumeMigrationSpec struct {
+	MigratedVolume []MigratedVolume `json:"migratedVolume,omitempty"`
+}
+
+type VolumeMigrationPhase string
+
+const (
+	VolumeMigrationPhasePending    = "Pending"
+	VolumeMigrationPhaseRunning    = "Running"
+	VolumeMigrationPhaseScheduling = "Scheduling"
+	VolumeMigrationPhaseSucceeded  = "Succeeded"
+	VolumeMigrationPhaseFailed     = "Failed"
+	VolumeMigrationPhaseUnknown    = "Unknown"
+)
+
+type VolumeMigrationPhaseTransitionTimestamp struct {
+	Phase                    VolumeMigrationPhase `json:"phase,omitempty"`
+	PhaseTransitionTimestamp metav1.Time          `json:"phaseTransitionTimestamp,omitempty"`
+}
+
+type MigratedVolumeState string
+
+const (
+        MigratedVolumeStateValid    = "Valid"
+        MigratedVolumeStatePending  = "Pending"
+        MigratedVolumeStateRejected = "Rejected"
+)
+
+type VolumeMigrationState struct {
+        SourcePvc string              `json:"sourcePvc,omitempty"`
+        State     MigratedVolumeState `json:"state,omitempty"`
+        Reason    *string             `json:"reason,omitempty"`
+}
+
+type VolumeMigrationStatus struct {
+        VolumeMigrationStates       []VolumeMigrationState                    `json:"volumeMigrationStates,omitempty"`
+        VirtualMachineInstanceName  *string                                   `json:"virtualMachineInstanceName,omitempty"`
+        VirtualMachineMigrationName *string                                    json:"virtualMachineMigrationName,omitempty"`
+        StartTimestamp              *metav1.Time                              `json:"startTimestamp,omitempty"`
+        EndTimestamp                *metav1.Time                              `json:"endTimestamp,omitempty"`
+        Phase                       VolumeMigrationPhase                      `json:"phase,omitempty"`
+        PhaseTransitionTimestamps   []VolumeMigrationPhaseTransitionTimestamp `json:"phaseTransitionTimestamps,omitempty"`
+}
+```
+
+Users may define more than one migrated volume, and those volumes may be associated with the same VMI, a different VMI, or none at all. The volumes defined on the same VMI will be categorized and grouped by the volume migration controller.
+
+At the moment, PVCs can only be migrated using this API if they belong to a single running virtual machine. Consequently, all volume migration objects that have migrated volumes that are associated with different VMIs will be rejected by the controller. The object is in the `Pending` state if none of the migrated volumes have been assigned to a VMI, since the VMI might be created at a later time.
+
+The `VolumeMigration` expects a list of source and destination PVCs and the destination PVCs needs to be created either manually by the user or other tools.
+
+According to the `ReclaimPolicySourcePvc`, KubeVirt will determine how to handle the source PVCs. The source PVC will be automatically deleted under the `Delete` policy, but it will remain untouched under the `Retain` policy.
```

Comment:
Shall we specify a default one if no policy is provided by a user? Maybe `Retain`?

Reply:
I don't know, probably, even if I'm not sure which default is the best. I was thinking more to delete policy as the default one.

Reply:
TBH, I do not have a strong opinion here. I checked `PersistentVolumeReclaimPolicy` and it has different defaults: `Retain` for manually created PVs and `Delete` for dynamic ones. So in our case, this could be `Delete` as well.

Reply:
yeah, I don't have a strong opinion either.  I guess we just need to well document what the default behavior is. Then, let's keep Delete as it is the case for PVCs